yarn-project/validator-client/src/validator.ts

```diff
@@ -55,15 +64,47 @@ export class ValidatorClient implements Validator {
 
   public registerBlockProposalHandler() {
     const handler = (block: BlockProposal): Promise<BlockAttestation> => {
-      return this.validationService.attestToProposal(block);
+      return this.attestToProposal(block);
     };
     this.p2pClient.registerBlockProposalHandler(handler);
   }
 
-  attestToProposal(proposal: BlockProposal) {
+  async attestToProposal(proposal: BlockProposal): Promise<BlockAttestation> {
+    // Check that all of the tranasctions in the proposal are available in the tx pool before attesting
+    await this.ensureTransactionsAreAvailable(proposal);
+    this.log.debug(`Transactions available, attesting to proposal with ${proposal.txs.length} transactions`);
+
+    // If the above function does not throw an error, then we can attest to the proposal
     return this.validationService.attestToProposal(proposal);
   }
 
+  /**
+   * Ensure that all of the transactions in the proposal are available in the tx pool before attesting
+   *
+   * 1. Check if the local tx pool contains all of the transactions in the proposal
+   * 2. If any transactions are not in the local tx pool, request them from the network
+   * 3. If we cannot retrieve them from the network, throw an error
+   * @param proposal - The proposal to attest to
+   */
+  async ensureTransactionsAreAvailable(proposal: BlockProposal) {
+    const txHashes: TxHash[] = proposal.txs;
+    const transactionStatuses = await Promise.all(txHashes.map(txHash => this.p2pClient.getTxStatus(txHash)));
+
+    const missingTxs = txHashes.filter((_, index) => !['pending', 'mined'].includes(transactionStatuses[index] ?? ''));
+
+    if (missingTxs.length === 0) {
+      return; // All transactions are available
+    }
+
+    this.log.verbose(`Missing ${missingTxs.length} transactions in the tx pool, requesting from the network`);
+
+    const requestedTxs = await this.p2pClient.requestTxs(missingTxs);
```

Comment:
For this check, should we have the check for for status further down? Just that right would we not end up requesting if we ever get the status wrong etc? 🤷 Just that even if used poorly would not want to request the things we already got.

Reply:
Something I got thinking about. As I read the code currently, if some other transaction is returned instead of the `missingTxs` would we not end up still signing here?

Reply:
1. Can we use anything other than the status check?
2. Validation checks to be added